### PR TITLE
[RDY] sort_heap algorithm implementation

### PR DESCRIPTION
**Pull request by ShivaniSD123**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -236,10 +236,10 @@ required description would be provided.
 
 ### Heap Operations
 
+- [x] sort_heap -> sort_heap, sort_heap_by
```
Comment (RishabhRD) [inline README.md:239]: Put sort_heap on the correct position.

File: src/algo/heap.rs [hunk 2]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
     pop_heap_by(rng, start, end, |x, y| x < y);
 }
 
+/// Converts heap in range into sorted range wrt cmp.
```
Comment (RishabhRD) [inline src/algo/heap.rs:365]: Converts given heap into sorted range wrt cmp.

File: src/algo/heap.rs [hunk 3]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+///
+/// # Precondition
+///  - `[start, end)` represents valid positions in rng.
+///  - rng at `[start, end)` is a heap.
+///
+/// # Postcondition
+///  - Sorts the elements in rng such that the whole range is in non-decending order.
```
Comment (RishabhRD) [inline src/algo/heap.rs:372]: sorts the element in rng at `[start, end)` such that whole range at `[start, end)` is in non-decreasing order wrt cmp.

File: src/algo/heap.rs [hunk 4]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+///  - Complexity: O(n.log2(n)) comparisions.
+///
```
Comment (RishabhRD) [inline src/algo/heap.rs:374]: Where n is number of elements in `[start, end)`.

See above.

File: src/algo/heap.rs [hunk 5]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+    cmp: Compare,
+) where
+    Range: RandomAccessRange + SemiOutputRange + ?Sized,
+    Compare: Fn(&Range::Element, &Range::Element) -> bool + Clone,
+{
+    while start != end {
+        pop_heap_by(rng, start.clone(), end.clone(), cmp.clone());
+        end = rng.before(end);
+    }
+}
```
Comment (RishabhRD) [inline src/algo/heap.rs:399]: Add a line break.

File: src/algo/heap.rs [hunk 6]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+/// Converts heap in range into sorted range.
```
Comment (RishabhRD) [inline src/algo/heap.rs:400]: Converts given heap into sorted range.

File: src/algo/heap.rs [hunk 7]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+///
+/// # Precondition
+///  - `[start, end)` represents valid positions in rng.
+///  - rng at `[start, end)` is a heap.
+///  - cmp should follow strict-weak-ordering relationship.
```
Comment (RishabhRD) [inline src/algo/heap.rs:405]: Delete cmp things from this whole comment

File: src/algo/heap.rs [hunk 8]
```diff
@@ -362,6 +362,74 @@ pub fn pop_heap<Range>(
...
+///
+/// # Postcondition
+///  - Sorts the elements in rng such that the whole range is in non-decending order wrt cmp.
+///  - Complexity: O(n.log2(n)) comparisions.
+///
```
Comment (RishabhRD) [inline src/algo/heap.rs:410]: Again, where n is ....

File: src/rng/heap.rs [hunk 9]
```diff
@@ -304,6 +304,81 @@ where
     algo::pop_heap(rng, start, end);
 }
 
+/// Converts heap in range into sorted range wrt cmp.
```
Comment (RishabhRD) [inline src/rng/heap.rs:307]: Converts given heap into sorted range wrt cmp.

File: src/rng/heap.rs [hunk 10]
```diff
@@ -304,6 +304,81 @@ where
...
+///
+/// # Precondition
+///  - rng is a heap.
+///
+/// # Postcondition
+///  - Sorts the elements in rng such that the whole range is in non-decending order.
```
Comment (RishabhRD) [inline src/rng/heap.rs:313]: wrt cmp.

File: src/rng/heap.rs [hunk 11]
```diff
@@ -304,6 +304,81 @@ where
...
+///  - Complexity: O(n.log2(n)) comparisions.
+///
```
Comment (RishabhRD) [inline src/rng/heap.rs:315]: Where n is number of elements in rng.

See above.

File: src/rng/heap.rs [hunk 12]
```diff
@@ -304,6 +304,81 @@ where
...
+/// use rng::infix::*;
+///
+/// let mut arr = [8, 7, 2, 5];
+/// let start = arr.start();
+/// let end = arr.end();
+/// rng::sort_heap_by(&mut arr, |x, y| x < y);
+/// assert!(arr.equals(&[2, 5, 7, 8]));
+///
+/// let mut arr = [9, 8, 7];
+/// arr.sort_heap();
```
Comment (RishabhRD) [inline src/rng/heap.rs:331]: its a sort_heap_by function documentation use sort_heap_by

File: src/rng/heap.rs [hunk 13]
```diff
@@ -304,6 +304,81 @@ where
...
+where
+    Range: RandomAccessRange + SemiOutputRange + ?Sized,
+    Compare: Fn(&Range::Element, &Range::Element) -> bool + Clone,
+{
+    let start = rng.start();
+    let end = rng.end();
+    algo::sort_heap_by(rng, start, end, cmp);
+}
+
+/// Converts heap in range into sorted range.
```
Comment (RishabhRD) [inline src/rng/heap.rs:344]: Converts given heap

File: src/rng/heap.rs [hunk 14]
```diff
@@ -304,6 +304,81 @@ where
...
+///
+/// # Precondition
+///   - rng is a heap.
+///   - cmp should follow strict-weak-ordering relationship.
```
Comment (RishabhRD) [inline src/rng/heap.rs:348]: Delete cmp things from this comment

File: src/rng/heap.rs [hunk 15]
```diff
@@ -304,6 +304,81 @@ where
...
+/// #### Infix Supported
+/// YES
+///  
+/// Example
+/// ```rust
+/// use stl::*;
+/// use rng::infix::*;
+///
+/// let mut arr = [8, 7, 2, 5];
+/// let start = arr.start();
```
Comment (RishabhRD) [inline src/rng/heap.rs:363]: remove start and end

File: src/rng/heap.rs [hunk 16]
```diff
@@ -304,6 +304,81 @@ where
...
+/// #### Infix Supported
+/// YES
+///
+/// Example
+/// ```rust
+/// use stl::*;
+/// use rng::infix::*;
+///
+/// let mut arr = [8, 7, 2, 5];
+/// let start = arr.start();
```
Comment (RishabhRD) [inline src/rng/heap.rs:325]: remove start and end.

**Review by RishabhRD — CHANGES_REQUESTED** [review]
Looking good to me, some doc changes required. Then we can merge it.

**Comment by ShivaniSD123** [comment]
Resolved the comments.

**Review by RishabhRD — APPROVED** [review]
LGTM!!!

**Comment by RishabhRD** [comment]
Thank You Shivani for your valuable contribution. If anytime you feel to contribute any other algorithm, feel free to do the same.

**Pull request merged**